### DISCUSSION TITLE
asio coroutines with stdexec

### DISCUSSION THREAD

**Discussion by mathisloge** [NVIDIA/stdexec#1237] [Q&A] [answered]
I know that there is https://github.com/NVIDIA/stdexec/discussions/1133 already, but it wasn't really answered. 

I was wondering if someone knows a way to combine stdexec's concepts and asio's `io_context`. 
Main reason is to allow to use stdexec but still use the existing ecosystem around asio. E.g. boost process v2 and so on. 

e.g. could a sender be created from an asio::awaitable? I'm quite new to stdexec and trying to get my mental model in a place where I can use it in any form. So I'm sorry if my question is on a quite high level, but I would also be happy if someone could just give me an opinion of whether this could work and whether I still have to wait for [P2855](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2855r0.html).

**Comment by maikel** [comment]
Do you ask for an asio completion token that makes asio functions return a sender? Iirc asio at least implemented *some* support for S&R (and deprecated it) but maintaining such a thing is difficult from the perspective of asio, because stdexec is a moving target.

I think its a good suggestion to have something like this in this repo

**Comment by maikel** [comment]
I had some time and made a CompletionToken for asio here https://godbolt.org/z/61jznd7x7

**Reply by mathisloge** [reply]
Oh wow, thank you for putting your time into this!! 
It will try it this weekend!

**Comment by SamGaaWaa** [comment, marked as answer]
Try this https://github.com/SamGaaWaa/asio2exec

**Reply by maikel** [reply]
How does cancelation work? Have you modified the asio code or is there an asio way to do that?

**Reply by SamGaaWaa** [reply]
> How does cancelation work? Have you modified the asio code or is there an asio way to do that?

I put the cancellation signal into stop callback and check the error code whether is equal to "operation_cancelled".

**Reply by maikel** [reply]
Can you cancel an outstanding read operation without canceling the whole socket?

**Reply by SamGaaWaa** [reply]
> Can you cancel an outstanding read operation without canceling the whole socket?

Yes, since asio support per-operation cancellation. https://www.boost.org/doc/libs/1_84_0/doc/html/boost_asio/overview/core/cancellation.html

**Reply by maikel** [reply]
Great that is the information that I needed. Thank you

**Reply by maikel** [reply]
I've updated my example here with cancellation. But, for me, `cancellation_signal/slot` is still a bit weird. I think I need to synchronize something between threads since it seems to be completely unsynchronized, right?

https://godbolt.org/z/7hPc3ne1s

**Reply by maikel** [reply]
I think there is indeed a race between registering a cancelation handler at initiation and a possible stop request from the receiver 😯

**Reply by SamGaaWaa** [reply]
Yes, see this example https://github.com/NVIDIA/stdexec/blob/main/test/test_common/senders.hpp

**Reply by mathisloge** [reply]
I've tested the header of @SamGaaWaa . Thanks. 
Since I was just beginning with my new project I've ported most of the async code from asio to stdexec. 
Just some asio::*_pipes are left and it connects quite well. 

@SamGaaWaa  maybe you could add a [license file](https://docs.github.com/de/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository#determining-the-location-of-your-license) to your repo.

**Reply by SamGaaWaa** [reply]
@mathisloge Ok.

**Comment by ericniebler** [comment]
I would be _thrilled_ to merge ASIO integration into stdexec.

**Reply by pfeatherstone** [reply]
It would be good to see a roadmap for networking in senders. After all, senders was born out of a dissatisfaction with Asio's executor's model and the committee presumably thought a better asynchrony model would suit networking. Is that right? I would have thought networking would be the #1 application for senders, ahead of cpu-bound tasking and GPUs.

**Comment by MoFHeka** [comment]
Same request. There is a lot of code based on the asio ecosystem, such as asynchronous Redis client. I really don't want to reinvent the wheel myself.
